File: day13-2.py
#!/usr/bin/env python3

import sys

def ints(itr):
    return [int(i) for i in itr]

with open(sys.argv[1], 'r') as f:
    file_lines = [l for l in f.read().strip().split('\n')]


in_nums = []

total = 0
result = 0
other = 0

SIZE = 2000
grid = [[False] * SIZE for i in range(SIZE)]
folds = []
while True:
    for l in file_lines:
        if not l:
            continue
        elif l.startswith('fold'):
            _, _, ins = l.split()
            axis, pos = ins.split('=')
            pos = int(pos)
            folds += [(axis, pos)]
        else:
            x, y = map(int, l.split(','))
            grid[y][x] = True

        if False:
            total += 1

    break

def print_grid():
    for r in grid:
        print(''.join(('#' if c else '.') for c in r))

width = SIZE
height = SIZE
# print_grid()
print(folds)
for axis, pos in folds:
    print(axis, pos)
    if axis == 'x':
        for x2 in range(pos + 1, width):
            for y in range(height):
                if not grid[y][x2]:
                    continue
                refp = pos - (x2 - (pos))
                grid[y][refp] = True
                grid[y][x2] = False
                width = pos + 1
    elif axis == 'y':
        for y2 in range(pos + 1, height):
            for x in range(width):
                if not grid[y2][x]:
                    continue
                refp = pos - (y2 - (pos))
                # print(x, y2)
                # print('refp', refp)
                grid[refp][x] = True
                grid[y2][x] = False
                height = pos + 1
    # print_grid()

for r in grid[:height]:
    print(''.join(('#' if c else '.') for c in r[:width]))

print(f'Total: {total}')
print(f'Result: {result}')
print(f'Other: {other}')

File: day13-2.cpp
#include "bits/stdc++.h"

using namespace std;

static void print_grid(int width, int height, const vector<pair<int, int>>& points) {
    vector<vector<bool>> grid(height);
    for (auto& row : grid) {
        row.resize(width);
    }
    for (const auto& [x, y] : points) {
        grid[y][x] = true;
    }

    for (int y = 0; y < height; y++) {
        for (int x = 0; x < width; x++) {
            cout << (grid[y][x] ? '#' : ' ');
        }
        cout << '\n';
    }
}

int main(int argc, char **argv) {
    freopen(argv[1], "r", stdin);

    vector<pair<int, int>> points;
    vector<pair<char, int>> folds;

    int width = 0;
    int height = 0;
    while (!cin.eof()) {
        string line;
        cin >> line;

        if (line.empty()) {
            continue;
        } else if (line == "fold") {
            cin >> line >> line;
            auto axis = line[0];

            auto pos = stoi(line.substr(2));
            folds.emplace_back(axis, pos);
        } else {
            auto x = stoi(line.substr(0, line.find(",")));
            auto y = stoi(line.substr(line.find(",") + 1, line.length()));
            points.emplace_back(x, y);

            if (x + 1 > width) {
                width = x + 1;
            }
            if (y + 1 > height) {
                height = y + 1;
            }
        }
    }

    for (const auto& [axis, fold_pos] : folds) {
        vector<pair<int, int>> new_points;
        new_points.reserve(points.size());

        if (axis == 'x') {
            for (auto [x, y] : points) {
                if (x > fold_pos) {
                    x = fold_pos - (x - fold_pos);
                    width = fold_pos + 1;
                }

                new_points.emplace_back(x, y);
            }
        } else {
            for (auto [x, y] : points) {
                if (y > fold_pos) {
                    y = fold_pos - (y - fold_pos);
                    height = fold_pos + 1;
                }

                new_points.emplace_back(x, y);
            }
        }

        points = std::move(new_points);
    }

    print_grid(width, height, points);

    return 0;
}
